docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,2624 @@
+At the time of writing of this document, TSC is still not fully mature as is evident from the above findings. It is therefore the current recommendation to not use it for any scheduling decisions.
+
+## Gardener API Changes
+
+### Shoot Spec
+
+We propose to extend the `shoot` resource to have a dedicated field under `spec.kubernetes.controlplane` which allows the consumer to define `highAvailability` configuration.
+
+There are a few options that were considered:
+
+> NOTE: In the below mentioned HA configuration, focus is on control plane HA. It is not clear on what would it mean to define HA configuration for data plane components. However, for better clarity of scope and to provide extensibility in future HA configurations are housed under `controlPlane`.
+
+**Option #1:** 
+```yaml
+kind: Shoot
+apiVersion: core.gardener.cloud/v1beta1
+spec:
+  kubernetes:
+    controlPlane:
+      highAvailability:
+        enabled: true # by default HA is disabled.
+        type: <single-zone | multi-zone>
+```
+In this option an existing `spec.kubernetes` nested field is used as a parent. Since `spec.kubernetes` hosts resources across control plane and data plane it is required to introduce another parent `controlPlane` under which `highAvailability` configuration can be defined.
+
+
+**Option #2**
+```yaml
+kind: Shoot
+apiVersion: core.gardener.cloud/v1beta1
+spec:
+  highAvailability:
+    controlPlane:
+      enabled: true # by default HA is disabled.
+      type: <single-zone | multi-zone>
+```
+In this option `highAvailability` is directly defined under `spec`. To ensure clarity of scope for high availability `controlPlane` is defined as a parent for hosting HA configuration.
+
+**Option #3**
+```yaml
+kind: Shoot
+apiVersion: core.gardener.cloud/v1beta1
+spec:
+  kubernetes:
+    controlPlane:
+      highAvailability:
+        enabled: true # by default HA is disabled.
+        failureDomain: 
+          type: <single-zone | multi-zone>
+          whenUnsatisfied: <DoNotSchedule | ScheduleAnyway>
+```
+This option is a variant of `Option-#1`. It provides an additional capability to the consumer to clearly define the behavior if the chosen type of failure domain cannot be provisioned by gardener due lack of available seed capable of either hosting a `single-zone` or `multi-zone` shoot control plane.
+
+**Option #4**
+```yaml
+kind: Shoot
+apiVersion: core.gardener.cloud/v1beta1
+spec:
+  kubernetes:
+    kubeAPIServer:
+      highAvailability:
+        enabled: true
+        type: <single-zone | multi-zone>
+    etcd:
+      highAvailability:
+        enabled: true
+        type: <single-zone | multi-zone>
+```
+
+This option takes the approach of specifying HA configuration at the component level. This is the most flexible option where one could choose which components should be setup in a HA mode. However this also comes with additional implementation complexity especially w.r.t validations.
+
```

Comment:
My favorite would be #3. Should the GEP not specify a recommendation or will that follow later?

Cosmetic: I would swap option #1 with #2 as that order would improve readability.